### PR TITLE
Block message version 2 (hash & validation)

### LINKED ISSUE #3788: Block hashing doesn't include justifications
## Overview

Block integrity is ensured by hashing of block data and signing the hash by the validator which produce a block.

The problem with [current (v0.12.6) implementation of block hashing](https://github.com/rchain/rchain/blob/v0.12.6/casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala#L286-L295) is that doesn't include justification field as part of hashed data.

```scala
def hashBlock(blockMessage: BlockMessage): BlockHash =
  ProtoUtil.hashByteArrays(
    blockMessage.header.toProto.toByteArray,
    blockMessage.body.toProto.toByteArray,
    blockMessage.sender.toByteArray,
    StringValue.of(blockMessage.sigAlgorithm).toByteArray,
    Int32Value.of(blockMessage.seqNum).toByteArray,
    StringValue.of(blockMessage.shardId).toByteArray,
    blockMessage.extraBytes.toByteArray
  )
```

It seems that the problem has roots even in the very first commit where input for block signature is calculated:
`hash(justification) + blockHash` :question:.
https://github.com/rchain/rchain/blob/b1e7c9a4f2d2da4b53b58d04452ea77d6d0e2439/casper/src/main/protobuf/coop/rchain/casper/protocol/CasperMessage.proto#L87-L93

Because of the sensitivity of the bug, this issue is created retroactively after fix PR is merged and a new release published on main net.
- https://github.com/rchain/rchain/pull/3693

## Consequence

This bug enables attackers to use original block proposed by the validator to make changes to justifications and send it to other validators effectively causing block creator to be slashed, probably choosing from a few slashing reasons like justification regression or invalid sequence number.

### PR DESCRIPTION
**Pull request by tgrospic**
## Overview

This PR fixes the major bug in block hashing function. To support compatibility to replay previous blocks, block version field is used mark _version 1_ with existing hashing and _version 2_ with the fix.

Details about the bug are specified in the linked issue.

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: casper/src/main/scala/coop/rchain/casper/Validate.scala [hunk 1]
```diff
@@ -165,8 +167,6 @@ object Validate {
       expirationThreshold: Int
   ): F[ValidBlockProcessing] =
     (for {
-      _ <- EitherT.liftF(Span[F].mark("before-block-hash-validation"))
-      _ <- EitherT(Validate.blockHash(block))
```
Comment (tgrospic) [inline casper/src/main/scala/coop/rchain/casper/Validate.scala:169]: Wrong block hash is now ignored.

File: integration-tests/test/test_slash.py [hunk 2]
```diff
@@ -32,6 +33,34 @@
 BONDED_VALIDATOR_KEY_3 = PrivateKey.from_hex("af47862137d4e772f540029ae73ee01443c61288f3df9307a13d681de6ad2de4")
 
 
+def blake2b_32(data: bytes = b'') -> hashlib.blake2b:
+  return hashlib.blake2b(data, digest_size=32)
+
+def gen_block_hash_from_block(block: BlockMessage) -> bytes:
```
Comment (tgrospic) [inline integration-tests/test/test_slash.py:40]: @zsluedem This function can be removed from PyRChain or kept in sync for every node version.

**Comment by tgrospic** [comment]
bors try

**Comment by tgrospic** [comment]
bors try

File: casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala [hunk 3]
```diff
@@ -293,6 +305,19 @@ object ProtoUtil {
...
     )
+
+  /**
+    * Create hash of a BlockMessage, all fields must be included except signature
+    */
+  def hashBlockVersion2(blockMessage: BlockMessage): BlockHash = {
+    val emptyBytes = ByteString.EMPTY
+    val blockClearSigData = blockMessage.copy(
+      blockHash = emptyBytes,
+      sig = emptyBytes
```
Comment (nzpr) [inline casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala:316]: Sig algorithm is still inside signed body? Did we want to make it out? I'm ok if we do this later, but maybe now is a good time.

File: casper/src/test/scala/coop/rchain/casper/util/BlockHashSignatureSpec.scala [hunk 4]
```diff
@@ -2,25 +2,74 @@ package coop.rchain.casper.util
...
+        val hashModified  = ProtoUtil.hashBlock(blockModified)
 
-      hashModified shouldBe hash
+        // Check hashes are different for modified block
+        hashModified !== hash ?= isModified
+      }
     }
   }
 
+  "block hash" should "not match for changed sequence number" in {
```
Comment (nzpr) [inline casper/src/test/scala/coop/rchain/casper/util/BlockHashSignatureSpec.scala:29]: Do we need more of these for all fields in block?

File: casper/src/main/scala/coop/rchain/casper/Validate.scala (hunk 1 above)
Comment (nzpr) [reply casper/src/main/scala/coop/rchain/casper/Validate.scala:169]: This should be done before saving to the block store, right?

**Review by nzpr — APPROVED** [review]
LGTM, minor comments

File: casper/src/main/scala/coop/rchain/casper/Validate.scala (hunk 1 above)
Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/Validate.scala:169]: Yes, block hash is checked in `checkIfWellFormedAndStore` together with the signature.

File: casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala (hunk 3 above)
Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala:316]: We can do it as part of 0.13 version. I was trying to keep minimal changes here.

File: casper/src/test/scala/coop/rchain/casper/util/BlockHashSignatureSpec.scala (hunk 4 above)
Comment (tgrospic) [reply casper/src/test/scala/coop/rchain/casper/util/BlockHashSignatureSpec.scala:29]: I'd like to have it. It's basically the same pattern for each field so good candidate for some macro to simplify.

**Pull request merged**